File: cpp/samples/types/decimal.cpp
/**
 * @file decimal.cpp
 * @brief Demonstrates Decimal usage with maritime vessel measurement values
 * @details This sample shows how to use the Vista SDK Decimal type for exact
 *          sensor readings, unit conversions, and calculated metrics typical
 *          of ship monitoring systems (propulsion, fuel, navigation, stability).
 */

#include <dnv/VistaSDK.h>

#include <iomanip>
#include <iostream>
#include <string>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk Decimal type samples ===\n\n";

    {
        std::cout << "1. Propulsion: Main engine temperatures and pressures\n";
        std::cout << "------------------------------------------------------\n";

        // VIS: 411.1/C101 - Main engine cooling water outlet temperature [°C]
        Decimal coolingWaterOutlet{ "87.3" };
        // VIS: 411.1/C102 - Main engine cooling water inlet temperature [°C]
        Decimal coolingWaterInlet{ "72.1" };
        auto deltaT = coolingWaterOutlet - coolingWaterInlet;

        // VIS: 411.1/C201 - Scavenge air pressure [bar]
        Decimal scavengeAirPressure{ "2.847" };

        // VIS: 411.1/C301 - Exhaust gas temperature after turbocharger [°C]
        Decimal exhaustGasTemp{ "342.6" };

        std::cout << "Cooling water outlet:  " << coolingWaterOutlet << " °C\n";
        std::cout << "Cooling water inlet:   " << coolingWaterInlet << " °C\n";
        std::cout << "Delta T (exact):       " << deltaT << " °C\n";
        std::cout << "Scavenge air pressure: " << scavengeAirPressure << " bar\n";
        std::cout << "Exhaust gas temp:      " << exhaustGasTemp << " °C\n";

        std::cout << "\n";
    }

    {
        std::cout << "2. Fuel: Consumption rate and voyage budget\n";
        std::cout << "---------------------------------------------\n";

        // VIS: 421/C101 - Main engine fuel oil consumption [t/day]
        Decimal fuelConsumptionPerDay{ "45.72" };
        // Voyage duration [days]
        Decimal voyageDays{ "12" };
        auto totalFuel = fuelConsumptionPerDay * voyageDays;

        // VIS: 421/C201 - Fuel oil density at 15°C [kg/m³]
        Decimal fuelDensity{ "0.8914" };
        // Convert total fuel [t] to volume [m³]  (t / density [t/m³])
        auto fuelVolumeM3 = totalFuel / fuelDensity;

        // Exact vs double - demonstrates why Decimal matters for fuel accounting
        double dConsumption{ 45.72 };
        double dDays{ 12.0 };
        double dResult = dConsumption * dDays;

        std::cout << "Fuel consumption:      " << fuelConsumptionPerDay << " t/day\n";
        std::cout << "Voyage duration:       " << voyageDays << " days\n";
        std::cout << "Total fuel (Decimal):  " << totalFuel << " t\n";
        std::cout << "Total fuel (double):   ";
        std::cout << std::fixed << std::setprecision(17) << dResult << "\n";
        std::cout << std::defaultfloat;
        std::cout << "Fuel volume:           " << fuelVolumeM3.round(3) << " m³\n";
        std::cout << "Density scale:         " << static_cast<int>(fuelDensity.scale()) << " decimal places\n";

        std::cout << "\n";
    }

    {
        std::cout << "3. Navigation: Speed, distance and ETA\n";
        std::cout << "-----------------------------------------\n";

        // VIS: 221/C101 - Speed over ground [knots]
        Decimal speedOverGround{ "14.3" };
        // VIS: 221/C201 - Speed through water [knots]
        Decimal speedThroughWater{ "13.8" };
        // Current contribution [knots]
        auto currentEffect = speedOverGround - speedThroughWater;

        // Remaining distance to waypoint [nm]
        Decimal distanceToWaypoint{ "1247.6" };
        // ETA [hours]
        auto etaHours = distanceToWaypoint / speedOverGround;
        // Convert to days and hours
        Decimal hoursPerDay{ "24" };
        auto etaDays = etaHours / hoursPerDay;

        std::cout << "Speed over ground:     " << speedOverGround << " kn\n";
        std::cout << "Speed through water:   " << speedThroughWater << " kn\n";
        std::cout << "Current effect:        " << currentEffect << " kn\n";
        std::cout << "Distance to waypoint:  " << distanceToWaypoint << " nm\n";
        std::cout << "ETA:                   " << etaHours.round(2) << " h  (" << etaDays.round(2) << " days)\n";

        std::cout << "\n";
    }

    {
        std::cout << "4. Stability: Draught, displacement and GM\n";
        std::cout << "--------------------------------------------\n";

        // VIS: 511/C101 - Forward draught [m]
        Decimal draughtFwd{ "7.843" };
        // VIS: 511/C102 - Aft draught [m]
        Decimal draughtAft{ "8.126" };
        // Mean draught [m]
        auto meanDraught = (draughtFwd + draughtAft) / Decimal{ "2" };
        // Trim [m] (positive = trimmed by stern)
        auto trim = draughtAft - draughtFwd;

        // Displacement [t] from hydrostatic tables (read from sensor/table at mean draught)
        Decimal displacement{ "23847.5" };
        // KG (centre of gravity above keel) [m]
        Decimal kg{ "9.214" };
        // KM (metacentre above keel) [m] from stability booklet
        Decimal km{ "10.871" };
        // GM (metacentric height) [m]
        auto gm = km - kg;

        std::cout << "Forward draught:       " << draughtFwd << " m\n";
        std::cout << "Aft draught:           " << draughtAft << " m\n";
        std::cout << "Mean draught:          " << meanDraught << " m\n";
        std::cout << "Trim:                  " << trim << " m (by stern)\n";
        std::cout << "Displacement:          " << displacement << " t\n";
        std::cout << "KG:                    " << kg << " m\n";
        std::cout << "KM:                    " << km << " m\n";
        std::cout << "GM (exact):            " << gm << " m\n";

        std::cout << "\n";
    }

    {
        std::cout << "5. Environmental: Celsius to Kelvin and unit conversions\n";
        std::cout << "----------------------------------------------------------\n";

        // Celsius to Kelvin offset - 273.15 is a known IEEE 754 float artifact trigger
        Decimal celsiusToKelvin{ "273.15" };

        // VIS: 412/C101 - Sea water temperature [°C]
        Decimal seaWaterTempC{ "14.8" };
        auto seaWaterTempK = seaWaterTempC + celsiusToKelvin;

        // VIS: 412/C102 - Outside air temperature [°C]
        Decimal airTempC{ "-3.2" };
        auto airTempK = celsiusToKelvin + airTempC;

        // Knots to m/s: 1 kn = 1852/3600 m/s
        Decimal knotsToMs{ "0.514444" };
        Decimal windSpeed{ "18.5" }; // [kn]  VIS: 412/C301 - Wind speed
        auto windSpeedMs = windSpeed * knotsToMs;

        // bar to kPa: 1 bar = 100 kPa
        Decimal barToKPa{ "100" };
        Decimal barometricPressure{ "1.0132" }; // [bar]  VIS: 412/C401
        auto pressureKPa = barometricPressure * barToKPa;

        // Demonstrate float artifact prevention: 273.15f would give "273.149994..."
        float celsiusOffsetFloat = 273.15f;
        Decimal fromFloat{ celsiusOffsetFloat };
        std::cout << std::boolalpha;
        std::cout << "Celsius -> Kelvin offset (string):  " << celsiusToKelvin << " K\n";
        std::cout << "From float 273.15f (exact?):       " << fromFloat << " K  "
                  << "(== \"273.15\": " << (fromFloat == Decimal{ "273.15" }) << ")\n";
        std::cout << "Sea water temp:    " << seaWaterTempC << " °C = " << seaWaterTempK << " K\n";
        std::cout << "Air temp:          " << airTempC << " °C = " << airTempK << " K\n";
        std::cout << "Wind speed:        " << windSpeed << " kn = " << windSpeedMs.round(3) << " m/s\n";
        std::cout << "Barometric press:  " << barometricPressure << " bar = " << pressureKPa.round(2) << " kPa\n";

        std::cout << "\n";
    }

    {
        std::cout << "6. Rounding modes: Sensor quantization and reporting\n";
        std::cout << "-------------------------------------------------------\n";

        // A raw sensor reading with many decimal places
        // VIS: 411.1/C401 - Shaft RPM [rpm]
        Decimal shaftRpm{ "97.4375" };

        std::cout << "Shaft RPM (raw):       " << shaftRpm << " rpm\n";
        std::cout << "Trunc  (floor abs):    " << shaftRpm.trunc() << " rpm\n";
        std::cout << "Floor  (toward -inf):  " << shaftRpm.floor() << " rpm\n";
        std::cout << "Ceil   (toward +inf):  " << shaftRpm.ceil() << " rpm\n";
        std::cout << "Round  1dp (banker's): " << shaftRpm.round(1) << " rpm\n";
        std::cout << "Round  1dp (ties-away):" << shaftRpm.round(1, Decimal::RoundingMode::ToNearestTiesAway)
                  << " rpm\n";

        // Accumulated flow meter - 1000 small increments of 0.047 t
        // VIS: 421/C301 - Fuel oil flow meter
        Decimal flowIncrement{ "0.047" };
        Decimal flowTotal{ "0" };
        for (int i = 0; i < 1000; ++i)
        {
            flowTotal += flowIncrement;
        }

        std::cout << "Flow total (1000 x 0.047 t): " << flowTotal << " t  "
                  << "(exact: " << std::boolalpha << (flowTotal == Decimal{ "47" }) << ")\n";

        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/types/decimal.rs
use vista_sdk::types::decimal::{Decimal, RoundingMode};

fn main() {
    println!("=== vista-sdk Decimal type samples ===\n");

    {
        println!("1. Propulsion: Main engine temperatures and pressures");
        println!("------------------------------------------------------");

        let cooling_water_outlet = Decimal::from_str("87.3").expect("valid");
        let cooling_water_inlet = Decimal::from_str("72.1").expect("valid");
        let delta_t = cooling_water_outlet - cooling_water_inlet;

        let scavenge_air_pressure = Decimal::from_str("2.847").expect("valid");
        let exhaust_gas_temp = Decimal::from_str("342.6").expect("valid");

        println!("Cooling water outlet:  {} °C", cooling_water_outlet);
        println!("Cooling water inlet:   {} °C", cooling_water_inlet);
        println!("Delta T (exact):       {} °C", delta_t);
        println!("Scavenge air pressure: {} bar", scavenge_air_pressure);
        println!("Exhaust gas temp:      {} °C", exhaust_gas_temp);
        println!();
    }

    {
        println!("2. Fuel: Consumption rate and voyage budget");
        println!("---------------------------------------------");

        let fuel_consumption_per_day = Decimal::from_str("45.72").expect("valid");
        let voyage_days = Decimal::from_i64(12);
        let total_fuel = fuel_consumption_per_day * voyage_days;

        let fuel_density = Decimal::from_str("0.8914").expect("valid");
        let fuel_volume_m3 = total_fuel / fuel_density;

        let d_consumption = 45.72_f64;
        let d_days = 12.0_f64;
        let d_result = d_consumption * d_days;

        println!("Fuel consumption:      {} t/day", fuel_consumption_per_day);
        println!("Voyage duration:       {} days", voyage_days);
        println!("Total fuel (Decimal):  {} t", total_fuel);
        println!("Total fuel (double):   {:.17}", d_result);
        println!(
            "Fuel volume:           {} m³",
            fuel_volume_m3.round(3, RoundingMode::ToNearest)
        );
        println!(
            "Density scale:         {} decimal places",
            fuel_density.scale()
        );
        println!();
    }

    {
        println!("3. Navigation: Speed, distance and ETA");
        println!("-----------------------------------------");

        let speed_over_ground = Decimal::from_str("14.3").expect("valid");
        let speed_through_water = Decimal::from_str("13.8").expect("valid");
        let current_effect = speed_over_ground - speed_through_water;

        let distance_to_waypoint = Decimal::from_str("1247.6").expect("valid");
        let eta_hours = distance_to_waypoint / speed_over_ground;
        let hours_per_day = Decimal::from_i64(24);
        let eta_days = eta_hours / hours_per_day;

        println!("Speed over ground:     {} kn", speed_over_ground);
        println!("Speed through water:   {} kn", speed_through_water);
        println!("Current effect:        {} kn", current_effect);
        println!("Distance to waypoint:  {} nm", distance_to_waypoint);
        println!(
            "ETA:                   {} h  ({} days)",
            eta_hours.round(2, RoundingMode::ToNearest),
            eta_days.round(2, RoundingMode::ToNearest)
        );
        println!();
    }

    {
        println!("4. Stability: Draught, displacement and GM");
        println!("--------------------------------------------");

        let draught_fwd = Decimal::from_str("7.843").expect("valid");
        let draught_aft = Decimal::from_str("8.126").expect("valid");
        let mean_draught = (draught_fwd + draught_aft) / Decimal::from_i64(2);
        let trim = draught_aft - draught_fwd;

        let displacement = Decimal::from_str("23847.5").expect("valid");
        let kg = Decimal::from_str("9.214").expect("valid");
        let km = Decimal::from_str("10.871").expect("valid");
        let gm = km - kg;

        println!("Forward draught:       {} m", draught_fwd);
        println!("Aft draught:           {} m", draught_aft);
        println!("Mean draught:          {} m", mean_draught);
        println!("Trim:                  {} m (by stern)", trim);
        println!("Displacement:          {} t", displacement);
        println!("KG:                    {} m", kg);
        println!("KM:                    {} m", km);
        println!("GM (exact):            {} m", gm);
        println!();
    }

    {
        println!("5. Environmental: Celsius to Kelvin and unit conversions");
        println!("----------------------------------------------------------");

        let celsius_to_kelvin = Decimal::from_str("273.15").expect("valid");

        let sea_water_temp_c = Decimal::from_str("14.8").expect("valid");
        let sea_water_temp_k = sea_water_temp_c + celsius_to_kelvin;

        let air_temp_c = Decimal::from_str("-3.2").expect("valid");
        let air_temp_k = celsius_to_kelvin + air_temp_c;

        let knots_to_ms = Decimal::from_str("0.514444").expect("valid");
        let wind_speed = Decimal::from_str("18.5").expect("valid");
        let wind_speed_ms = wind_speed * knots_to_ms;

        let bar_to_kpa = Decimal::from_i64(100);
        let barometric_pressure = Decimal::from_str("1.0132").expect("valid");
        let pressure_kpa = barometric_pressure * bar_to_kpa;

        let celsius_offset_float = 273.15_f32;
        let from_float = Decimal::from_f64(celsius_offset_float as f64);
        println!(
            "Celsius -> Kelvin offset (string):  {} K",
            celsius_to_kelvin
        );
        println!(
            "From float 273.15f (exact?):        {} K  (== \"273.15\": {})",
            from_float,
            from_float == celsius_to_kelvin
        );
        println!(
            "Sea water temp:    {} °C = {} K",
            sea_water_temp_c, sea_water_temp_k
        );
        println!("Air temp:          {} °C = {} K", air_temp_c, air_temp_k);
        println!(
            "Wind speed:        {} kn = {} m/s",
            wind_speed,
            wind_speed_ms.round(3, RoundingMode::ToNearest)
        );
        println!(
            "Barometric press:  {} bar = {} kPa",
            barometric_pressure,
            pressure_kpa.round(2, RoundingMode::ToNearest)
        );
        println!();
    }

    {
        println!("6. Rounding modes: Sensor quantization and reporting");
        println!("-------------------------------------------------------");

        let shaft_rpm = Decimal::from_str("97.4375").expect("valid");

        println!("Shaft RPM (raw):             {} rpm", shaft_rpm);
        println!("Trunc  (floor abs):          {} rpm", shaft_rpm.trunc());
        println!("Floor  (toward -inf):        {} rpm", shaft_rpm.floor());
        println!("Ceil   (toward +inf):        {} rpm", shaft_rpm.ceil());
        println!(
            "Round  1dp (banker's):       {} rpm",
            shaft_rpm.round(1, RoundingMode::ToNearest)
        );
        println!(
            "Round  1dp (ties-away):      {} rpm",
            shaft_rpm.round(1, RoundingMode::ToNearestTiesAway)
        );

        let tied_even = Decimal::from_str("97.45").expect("valid");
        let tied_odd = Decimal::from_str("97.35").expect("valid");

        println!();
        println!("Exact ties at 1dp - banker's rounds to even, ties-away up:");
        println!(
            "  {}  banker's: {}   ties-away: {}   (97.4 has even last digit, stays)",
            tied_even,
            tied_even.round(1, RoundingMode::ToNearest),
            tied_even.round(1, RoundingMode::ToNearestTiesAway)
        );
        println!(
            "  {}  banker's: {}   ties-away: {}   (97.3 is odd, up to 97.4)",
            tied_odd,
            tied_odd.round(1, RoundingMode::ToNearest),
            tied_odd.round(1, RoundingMode::ToNearestTiesAway)
        );

        let flow_increment = Decimal::from_str("0.047").expect("valid");
        let mut flow_total = Decimal::zero();
        for _ in 0..1000 {
            flow_total += flow_increment;
        }
        println!(
            "Flow total (1000 x 0.047 t): {} t  (exact: {})",
            flow_total,
            flow_total == Decimal::from_i64(47)
        );
        println!();
    }
}
